File: drivers/go/driver.go
// Crucible Go driver — persistent replay front-end for the differential loop.
//
// Speaks drivers/common/CONTRACT.md: reads length-prefixed records on stdin,
// decodes each into the probe message via the generated corelib-go code, and
// writes one canonical line per record (oracle/canonical.md) to stdout.
//
// The Go coverage engine is native fuzzing — see fuzz_test.go (FuzzProbe).
package main

import (
	"bufio"
	"encoding/binary"
	"encoding/hex"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"math"
	"os"
	"reflect"
	"strings"

	sofab "github.com/sofa-buffers/corelib-go"

	msg "crucible/driver/go/message"
)

// materialize (oracle/materialized.md), gated by SOFAB_MATERIALIZE=1. Instead of
// the round-trip hex, dump the DECODED value: every field, every array element made
// explicit. Floats are raw IEEE-754 bit patterns (math.Float32bits/Float64bits),
// never a decimal rendering.
//
// The walk is schema-AGNOSTIC: it holds no hardcoded field layout. The structure —
// field ids, names, kinds, array counts, nesting — is driven entirely by the
// generated descriptor (engine/structured/schema.py → oracle/materialized-schema.json),
// loaded at startup. The generated Go struct is navigated by reflection, resolving
// each descriptor node's schema name against the field's `json:"<schema-name>"` tag.
// Only the per-kind LEAF formatting below is schema-specific.
var materialized = os.Getenv("SOFAB_MATERIALIZE") == "1"

// schemaNode is one node of the materialized descriptor (oracle/materialized-schema.json).
// A leaf carries only kind; struct carries fields; array/wrapper carry elem + count.
type schemaNode struct {
	ID     int          `json:"id"`
	Name   string       `json:"name"`
	Kind   string       `json:"kind"`
	Elem   string       `json:"elem"`
	Count  int          `json:"count"`
	Fields []schemaNode `json:"fields"`
}

// schemaDoc is the top of the descriptor: { "message": ..., "fields": [node,...] }.
type schemaDoc struct {
	Message string       `json:"message"`
	Fields  []schemaNode `json:"fields"`
}

// schema is the parsed descriptor, populated at startup in materialize mode only.
var schema schemaDoc

// loadSchema reads the descriptor from $SOFAB_MATERIALIZE_SCHEMA (fallback
// oracle/materialized-schema.json) and parses it. Fatal on failure — a materialize
// run with no descriptor cannot produce a correct dump.
func loadSchema() {
	path := os.Getenv("SOFAB_MATERIALIZE_SCHEMA")
	if path == "" {
		path = "oracle/materialized-schema.json"
	}
	data, err := os.ReadFile(path)
	if err != nil {
		fmt.Fprintln(os.Stderr, "crucible-go: cannot read materialize schema:", err)
		os.Exit(1)
	}
	if err := json.Unmarshal(data, &schema); err != nil {
		fmt.Fprintln(os.Stderr, "crucible-go: cannot parse materialize schema:", err)
		os.Exit(1)
	}
}

func mF32(x float32) string { return fmt.Sprintf("f%08x", math.Float32bits(x)) }
func mF64(x float64) string { return fmt.Sprintf("F%016x", math.Float64bits(x)) }
func mText(s string) string {
	return fmt.Sprintf("t%d:%s", len(s), hex.EncodeToString([]byte(s)))
}
func mBlob(b []byte) string {
	return fmt.Sprintf("b%d:%s", len(b), hex.EncodeToString(b))
}

// mLeaf formats one scalar value per its schema kind (u|s|fp32|fp64|string|blob),
// pulling the concrete value out of the reflected field. This is the only
// schema-specific formatting left; every leaf and every array element flows here.
func mLeaf(kind string, v reflect.Value) string {
	switch kind {
	case "u":
		return fmt.Sprintf("u%d", v.Uint())
	case "s":
		return fmt.Sprintf("s%d", v.Int())
	case "fp32":
		return mF32(float32(v.Float()))
	case "fp64":
		return mF64(v.Float())
	case "string":
		return mText(v.String())
	case "blob":
		return mBlob(v.Bytes())
	}
	panic("materialize: unhandled leaf kind " + kind)
}

// mDefault is the materialized type default for a numeric/fp array element, used to
// fill-to-N when the Go corelib leaves an absent array nil (short of its schema count).
func mDefault(elem string) string {
	switch elem {
	case "u":
		return "u0"
	case "s":
		return "s0"
	case "fp32":
		return "f00000000"
	case "fp64":
		return "F0000000000000000"
	}
	panic("materialize: unhandled array elem " + elem)
}

// fieldByTag returns the struct field of v whose `json` tag equals the schema name.
// The generated struct exports PascalCase names but tags each with its schema name,
// so this is the schema-name → Go-field bridge that keeps the walk schema-agnostic.
func fieldByTag(v reflect.Value, name string) reflect.Value {
	t := v.Type()
	for i := 0; i < t.NumField(); i++ {
		tag := t.Field(i).Tag.Get("json")
		if i := strings.IndexByte(tag, ','); i >= 0 {
			tag = tag[:i]
		}
		if tag == name {
			return v.Field(i)
		}
	}
	panic("materialize: no struct field with json tag " + name)
}

// walk renders one descriptor node against its reflected value (oracle/materialized.md).
func walk(n *schemaNode, v reflect.Value) string {
	switch n.Kind {
	case "struct":
		parts := make([]string, len(n.Fields))
		for i := range n.Fields {
			c := &n.Fields[i]
			parts[i] = fmt.Sprintf("%d:%s", c.ID, walk(c, fieldByTag(v, c.Name)))
		}
		return "{" + strings.Join(parts, ";") + "}"
	case "array":
		// numeric/fp fixed-count array: emit in-memory elements, then fill-to-N.
		out := make([]string, 0, n.Count)
		for i := 0; i < v.Len(); i++ {
			out = append(out, mLeaf(n.Elem, v.Index(i)))
		}
		for len(out) < n.Count {
			out = append(out, mDefault(n.Elem))
		}
		return "[" + strings.Join(out, ",") + "]"
	case "wrapper":
		// string_array/blob_array: emit the container's actual elements in index
		// order (its length is itself the signal — no fill-to-N).
		out := make([]string, v.Len())
		for i := range out {
			out[i] = mLeaf(n.Elem, v.Index(i))
		}
		return "[" + strings.Join(out, ",") + "]"
	default:
		return mLeaf(n.Kind, v)
	}
}

// materialize walks the whole message: the descriptor's top `fields` list against
// the decoded msg.Probe value (oracle/materialized.md).
func materialize(m *msg.Probe) string {
	v := reflect.ValueOf(m).Elem()
	parts := make([]string, len(schema.Fields))
	for i := range schema.Fields {
		c := &schema.Fields[i]
		parts[i] = fmt.Sprintf("%d:%s", c.ID, walk(c, fieldByTag(v, c.Name)))
	}
	return "{" + strings.Join(parts, ";") + "}"
}

// canonical writes the canonical line for one candidate input
// (oracle/canonical.md: decode -> re-encode -> hex).
func canonical(w *bufio.Writer, data []byte) {
	m, err := msg.DecodeProbe(data)
	if err != nil {
		if errors.Is(err, sofab.ErrIncomplete) {
			// INCOMPLETE (MESSAGE_SPEC §7): decode ended mid-message — the third
			// canonical verdict, neither accept nor reject. The corelib returns no
			// partial value here (DecodeProbe drops it), so emit the bare `I`.
			fmt.Fprint(w, "I\n")
			return
		}
		if errors.Is(err, sofab.ErrLimitExceeded) {
			// LIMIT_EXCEEDED (generator#102, limit mode only): a configured receiver-side
			// cap on a schema-unbounded field. A policy rejection distinct from INVALID.
			fmt.Fprint(w, "L\n")
			return
		}
		// Coarse reject class (class comparison is soft; see oracle/policy.yaml).
		fmt.Fprint(w, "R invalid_msg\n")
		return
	}
	if materialized {
		// SOFAB_MATERIALIZE=1: dump the decoded value (oracle/materialized.md)
		// instead of the round-trip hex. Only the `A` payload changes.
		fmt.Fprintf(w, "A %s\n", materialize(m))
		return
	}
	b, err := m.Encode()
	if err != nil {
		fmt.Fprint(w, "R other\n")
		return
	}
	fmt.Fprintf(w, "A %s\n", hex.EncodeToString(b))
}

func main() {
	if materialized {
		loadSchema()
	}

	r := bufio.NewReader(os.Stdin)
	w := bufio.NewWriter(os.Stdout)
	defer w.Flush()

	var lenbuf [4]byte
	for {
		_, err := io.ReadFull(r, lenbuf[:])
		if err == io.EOF {
			return // clean EOF at record boundary
		}
		if err != nil {
			fmt.Fprintln(os.Stderr, "crucible-go: short length prefix:", err)
			os.Exit(1)
		}
		n := binary.LittleEndian.Uint32(lenbuf[:])
		data := make([]byte, n)
		if n > 0 {
			if _, err := io.ReadFull(r, data); err != nil {
				fmt.Fprintln(os.Stderr, "crucible-go: short payload:", err)
				os.Exit(1)
			}
		}
		canonical(w, data)
		w.Flush()
	}
}

File: drivers/cs/Driver.cs
// Crucible C# driver — persistent replay front-end for the differential loop.
//
// Speaks drivers/common/CONTRACT.md: reads length-prefixed records on stdin,
// decodes each into the probe message via the generated corelib-cs code, and
// writes one canonical line (oracle/canonical.md) per record to stdout.
//
// Verdict is three-valued (oracle/canonical.md, MESSAGE_SPEC §7): A COMPLETE /
// I INCOMPLETE / R INVALID. Single-pass decode via the generated status-returning
// `Probe.TryDecode(byte[], out Probe)` (sofabgen 0.16.0 — G-0008 fixed, see
// docs/SOFABGEN.md): it feeds the bytes AND returns the terminal `DecodeStatus`,
// so one call yields both the verdict (its returned status, or the SofabException
// it throws on malformed input) and the decoded value (`msg`, re-encoded for the
// hex on a COMPLETE decode). This replaces the earlier two-pass workaround that
// re-ran `IStream.Feed` against a null visitor because `Probe.Decode` discarded
// the status.
//
// The C# coverage engine is SharpFuzz — see Fuzz.cs.
using System;
using System.Collections.Generic;
using System.Globalization;
using System.IO;
using System.Text;
using System.Text.Json;
using sofab;
using Message;

namespace Crucible;

internal static class Driver
{
    // Materialized value dump (oracle/materialized.md), SOFAB_MATERIALIZE=1.
    // On a COMPLETE decode this walks the DECODED value and dumps every field +
    // every array element explicitly, instead of re-encoding to wire hex. The walk
    // is SCHEMA-AGNOSTIC: it is driven entirely by the generated descriptor
    // (engine/structured/schema.py → oracle/materialized-schema.json), reflecting
    // field values out of the generated Message.cs classes by name. Nothing about
    // the message's structure is hardcoded here — only the per-kind LEAF formatting
    // is (u/s/fp32/fp64/string/blob). The engine/structured/materialize.py reference
    // is ground truth; this output must equal it byte-for-byte. fp32 is printed
    // straight from its float bit pattern (the corelib decodes it as a native float,
    // so no widen/repack fidelity caveat applies here — contrast the Python driver).
    private static readonly bool Materialize =
        Environment.GetEnvironmentVariable("SOFAB_MATERIALIZE") == "1";

    // A descriptor node (recursive): the generated schema shape, loaded at startup.
    // Leaf: id/name/kind. struct: +fields. array: +elem(u|s|fp32|fp64)+count.
    // wrapper: +elem(string|blob)+count.
    private sealed class SchemaNode
    {
        public int Id;
        public string Name;
        public string Kind;
        public string Elem;
        public int Count;
        public List<SchemaNode> Fields;
    }

    // The generated message descriptor's top-level fields, loaded only in
    // materialize mode (parsing is skipped entirely on the default hex path).
    private static readonly List<SchemaNode> Schema = Materialize ? LoadSchema() : null;

    private static List<SchemaNode> LoadSchema()
    {
        string path = Environment.GetEnvironmentVariable("SOFAB_MATERIALIZE_SCHEMA")
            ?? "oracle/materialized-schema.json";
        using var doc = JsonDocument.Parse(File.ReadAllText(path));
        var fields = new List<SchemaNode>();
        foreach (var f in doc.RootElement.GetProperty("fields").EnumerateArray())
            fields.Add(ParseNode(f));
        return fields;
    }

    private static SchemaNode ParseNode(JsonElement e)
    {
        var n = new SchemaNode
        {
            Id = e.GetProperty("id").GetInt32(),
            Kind = e.GetProperty("kind").GetString(),
        };
        if (e.TryGetProperty("name", out var nm)) n.Name = nm.GetString();
        if (e.TryGetProperty("elem", out var el)) n.Elem = el.GetString();
        if (e.TryGetProperty("count", out var ct)) n.Count = ct.GetInt32();
        if (e.TryGetProperty("fields", out var fs))
        {
            n.Fields = new List<SchemaNode>();
            foreach (var c in fs.EnumerateArray()) n.Fields.Add(ParseNode(c));
        }
        return n;
    }

    private static string Hex(byte[] b)
    {
        var sb = new StringBuilder(b.Length * 2);
        foreach (byte x in b) sb.Append(x.ToString("x2", CultureInfo.InvariantCulture));
        return sb.ToString();
    }

    private static string U(ulong v) => "u" + v.ToString(CultureInfo.InvariantCulture);
    private static string S(long v) => "s" + v.ToString(CultureInfo.InvariantCulture);
    private static string F32(float x) =>
        "f" + ((uint)BitConverter.SingleToInt32Bits(x)).ToString("x8", CultureInfo.InvariantCulture);
    private static string F64(double x) =>
        "F" + ((ulong)BitConverter.DoubleToInt64Bits(x)).ToString("x16", CultureInfo.InvariantCulture);

    private static string T(string s)
    {
        byte[] b = Encoding.UTF8.GetBytes(s ?? "");
        return "t" + b.Length.ToString(CultureInfo.InvariantCulture) + ":" + Hex(b);
    }

    private static string B(byte[] bb)
    {
        bb ??= Array.Empty<byte>();
        return "b" + bb.Length.ToString(CultureInfo.InvariantCulture) + ":" + Hex(bb);
    }

    // Format one leaf value per its descriptor kind (oracle/materialized.md §Grammar).
    // The element declared type (u/s/fp32/fp64/string/blob) comes from the schema, so
    // array/wrapper elements reuse this exactly as scalar leaves do.
    private static string Leaf(string kind, object v) => kind switch
    {
        "u" => U(Convert.ToUInt64(v, CultureInfo.InvariantCulture)),
        "s" => S(Convert.ToInt64(v, CultureInfo.InvariantCulture)),
        "fp32" => F32(Convert.ToSingle(v, CultureInfo.InvariantCulture)),
        "fp64" => F64(Convert.ToDouble(v, CultureInfo.InvariantCulture)),
        "string" => T((string)v),
        "blob" => B((byte[])v),
        _ => throw new InvalidOperationException("unhandled leaf kind " + kind),
    };

    // Read a message field/property by its schema name via reflection. corelib-cs's
    // generated Message.cs names members with the schema's exact casing (snake_case,
    // e.g. bytes_field / string_array) and exposes them as public fields; a property
    // is accepted too, so this survives a codegen convention change.
    private static object GetMember(object obj, string name, out Type declared)
    {
        Type t = obj.GetType();
        var f = t.GetField(name);
        if (f != null) { declared = f.FieldType; return f.GetValue(obj); }
        var p = t.GetProperty(name);
        if (p != null) { declared = p.PropertyType; return p.GetValue(obj); }
        throw new MissingMemberException(t.Name, name);
    }

    // Generic, schema-driven walk of the decoded value. Structure comes entirely from
    // the descriptor node; only Leaf() is schema-specific (per-kind formatting).
    private static string Walk(SchemaNode node, object value)
    {
        switch (node.Kind)
        {
            case "struct":
                return WalkStruct(node.Fields, value);
            case "array":
            {
                // Fixed-count numeric/fp array, materialized to its full N in memory:
                // every element emitted, no trailing trim (that only elides on the wire).
                var arr = (Array)value;
                var sb = new StringBuilder("[");
                for (int i = 0; i < arr.Length; i++)
                {
                    if (i > 0) sb.Append(',');
                    sb.Append(Leaf(node.Elem, arr.GetValue(i)));
                }
                return sb.Append(']').ToString();
            }
            case "wrapper":
            {
                // string_array / blob_array: the decoded container is already grown to
                // highest-populated index + 1 with interior gaps as empty elements —
                // emit all Count elements in index order.
                var list = (System.Collections.IList)value;
                var sb = new StringBuilder("[");
                for (int i = 0; i < list.Count; i++)
                {
                    if (i > 0) sb.Append(',');
                    sb.Append(Leaf(node.Elem, list[i]));
                }
                return sb.Append(']').ToString();
            }
            default:
                return Leaf(node.Kind, value);
        }
    }

    private static string WalkStruct(List<SchemaNode> fields, object value)
    {
        var sb = new StringBuilder("{");
        for (int i = 0; i < fields.Count; i++)
        {
            SchemaNode c = fields[i];
            if (i > 0) sb.Append(';');
            object cv = GetMember(value, c.Name, out Type ct);
            // A nested struct member is initialized non-null by the generated code, but
            // guard defensively (mirrors the old `?? new` walk) so a null never NREs.
            if (c.Kind == "struct" && cv == null) cv = Activator.CreateInstance(ct);
            sb.Append(c.Id.ToString(CultureInfo.InvariantCulture)).Append(':').Append(Walk(c, cv));
        }
        return sb.Append('}').ToString();
    }

    // Dump the decoded value per oracle/materialized.md: a single-line object, fields
    // in ascending id order (the descriptor's order), no field ever omitted.
    private static string MaterializeValue(Probe m) => WalkStruct(Schema, m);

    private static string RejectClass(SofabException e) => e.Error switch
    {
        SofabError.InvalidMessage => "invalid_msg",
        SofabError.Argument => "argument",
        SofabError.Usage => "usage",
        SofabError.BufferFull => "buffer_full",
        _ => "other",
    };

    // Map a decode exception to its canonical line prefix. LIMIT_EXCEEDED
    // (generator#102, limit mode only) is a policy rejection distinct from INVALID
    // and gets its own verdict `L`; everything else is an `R <class>` reject.
    private static string ErrLine(SofabException e) =>
        e.Error == SofabError.LimitExceeded ? "L" : "R " + RejectClass(e);

    private static string Canonical(byte[] data)
    {
        // One pass: TryDecode fills `m` and returns the corelib's real §7 outcome
        // (or throws SofabException on malformed input).
        DecodeStatus status;
        Probe m;
        try
        {
            status = Probe.TryDecode(data, out m);
        }
        catch (SofabException e) { return ErrLine(e); }
        catch (Exception) { return "R other"; }

        // INCOMPLETE: bytes ended mid-message — the third verdict, not an error.
        // (Optional I <hex> partial-value payload is not materialized here.)
        if (status == DecodeStatus.Incomplete) return "I";

        // COMPLETE, materialize mode (oracle/materialized.md): dump the decoded
        // value's fields/elements explicitly instead of re-encoding to wire hex.
        if (Materialize) return "A " + MaterializeValue(m);

        // COMPLETE: value via re-encode -> hex.
        byte[] enc;
        try
        {
            enc = m.Encode();
        }
        catch (SofabException e) { return ErrLine(e); }
        catch (Exception) { return "R other"; }

        var sb = new StringBuilder("A ");
        foreach (byte b in enc) sb.Append(b.ToString("x2"));
        return sb.ToString();
    }

    private static bool ReadFully(Stream s, byte[] buf, int n)
    {
        int off = 0;
        while (off < n)
        {
            int r = s.Read(buf, off, n - off);
            if (r <= 0) return false;
            off += r;
        }
        return true;
    }

    private static void Main()
    {
        Stream stdin = Console.OpenStandardInput();
        var w = new StreamWriter(Console.OpenStandardOutput(), new UTF8Encoding(false));
        w.NewLine = "\n";
        byte[] lenbuf = new byte[4];
        while (true)
        {
            if (!ReadFully(stdin, lenbuf, 4)) break; // clean EOF at record boundary
            uint n = (uint)(lenbuf[0] | (lenbuf[1] << 8) | (lenbuf[2] << 16) | (lenbuf[3] << 24));
            byte[] data = new byte[n];
            if (n > 0 && !ReadFully(stdin, data, (int)n))
            {
                Console.Error.WriteLine("crucible-cs: short payload");
                Environment.Exit(1);
            }
            w.Write(Canonical(data));
            w.Write('\n');
            w.Flush();
        }
    }
}
